client/src/shared/component/CircleButton.tsx

```diff
@@ -0,0 +1,49 @@
+import * as React from "react";
+import { Slot } from "@radix-ui/react-slot";
+import { cva, type VariantProps } from "class-variance-authority";
+import { cn } from "@/lib/utils";
+
+const circleButtonVariants = cva(
+  "focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive inline-flex cursor-pointer items-center justify-center whitespace-nowrap transition-all outline-none focus-visible:ring-[3px] disabled:pointer-events-none disabled:bg-gray-200 disabled:bg-none disabled:text-gray-500 [&_svg]:shrink-0",
+  {
+    variants: {
+      variant: {
+        default:
+          "typo-body-4 gap-1.5 rounded-[var(--radius-50)] bg-white text-black shadow-main transition-all duration-200 hover:h-[4.5625rem] hover:w-[10.625rem] hover:translate-x-[30%] hover:pl-4 [&_span]:hidden hover:[&_span]:inline-flex [&_svg]:h-[2.1219rem] [&_svg]:w-[2.1219rem] hover:[&_svg]:h-[2.625rem] hover:[&_svg]:w-[2.625rem]",
+        gradient:
+          "typo-body-5 border-circle-gradient-main hover:[&_span]:text-body-5 bg-white text-black shadow-main transition-all duration-200 hover:h-[3.7763rem] hover:w-[11.3281rem] hover:translate-x-[30%] hover:border-none hover:bg-gradient-main-diagonal [&_span]:hidden hover:[&_span]:inline-flex hover:[&_span]:text-white [&_svg]:h-[1.25rem] [&_svg]:w-[1.4375rem] hover:[&_svg]:hidden",
+        ghost: "bg-transparent [&_svg]:h-[2.0625rem] [&_svg]:w-[2.375rem]",
+      },
+      size: {
+        default: "h-[3.75rem] w-[3.75rem]",
+      },
+    },
+    defaultVariants: {
+      variant: "default",
+      size: "default",
+    },
+  }
+);
+
+function CircleButton({
+  className,
+  variant,
+  size,
+  asChild = false,
+  ...props
+}: React.ComponentProps<"button"> &
+  VariantProps<typeof circleButtonVariants> & {
+    asChild?: boolean;
+  }) {
+  const Comp = asChild ? Slot : "button";
+
+  return (
+    <Comp
+      data-slot="button"
+      className={cn(circleButtonVariants({ variant, size, className }))}
+      {...props}
+    />
+  );
```

Comment:
p2: <span> 태그로 텍스트를 감싸지 않으면 hover가 정상적으로 동작하지 않는 구조라는 점에서 개발자의 실수를 유발할 수도 있을 것 같아요.!
text를 별도 props로 받고, 내부에서 <span> 태그로 감싸 렌더링하는 구조는 어떠신가요? 책임을 전부 CircleButton 내부로 위임할 수 있어서 좀 더 안전하게 사용할 수 있을 것 같습니다.


```suggestion
    <Comp
      data-slot="button"
      className={cn(circleButtonVariants({ variant, size }), className)}
      {...props}
    >
      {icon}
      {text && <span>{text}</span>}
    </Comp>
  );
```

```tsx
<CircleButton variant="default" icon={<EditIcon />} text="편집 하기" />
```

이런 방식으로 수정하는걸 고민해봤는데.. 어떻게 생각하시나요! 🤓

Reply:
저도 같은 고민으로,  circle button 내부에서 children이 string일 경우에만 span 태그로 감싸서 반환하게 하도록 시도했었는데 icon과 text가 잘 나타나지 않는 문제가 발생했었습니다.

제안해주신 방법처럼 children을 한꺼번에 받지 않고, props로 구분해서 받으니 circle button 내부에서 span 부분을 올바르게 처리할 수 있었습니다. 감사합니다! 🙏